docs: add commentary on new Flux design (#2722)
influxdata/flux: docs/NewFlux_Comentary.md

The current Flux internals use the interpreter to create DAG of promises to data to construct a query spec.
That spec is then planned etc and passed down to the execution engine.

So long as we can continue to build a query spec from this new Flux API, we can continue to execute the queries.
My rough plan is that a set of new Flux functions will be added to a `table` package.
The `table` package will be very similar to the current `universe` package. It will contain all basic functions and new functions as well like `aggregate`.

This new `table` package will need to provide its own mechanism for producing a query spec.
The current execution engine can handle streams of tables with heterogenous schemas.
A homogenous schema is a subset of functionality in a heterogeneous schema and as such can run on the execution engine without issue.
We will still leverage the engines concept of groups to do aggregates etc but those details will be hidden from the user.

As a rough example to illustrate how this would work, the following new Flux:

```
@data.temperatures{location,state,degrees_bottom}
 |> aggregate({min(degrees_bottom)}, by:["location"])
```

> would generate a spec similar to what this Flux would generate

Reviewer: ```suggestion
Would generate a spec similar to what this Flux would generate:
```
Consistent with the ccpitalization and colon above.

```
from(bucket:"data")

>  |> filter(fn:(r) => r._measurement == "temperatures")

Reviewer 2: There is a missing filter predicate `r._field == "degrees_bottom"`, right?  It seems like we would need to know which columns are InfluxDB tags and which are fields statically to make this transformation?

Reviewer: I'm confused. Do we need to filter on the field here in order to perform an aggregate function on it?

Reviewer 2: I guess since the `from` function will produce pivoted data, there will be no `_field` column.  But I still don't see how we specify which fields from a measurement to get from storage.  Is it all of then, by default?

Author: @wolffcm I was thinking we would send the entire list of columns down to storage and let it figure it out compared to the series index. It should be able to work out which columns are tags and which are fields at that point. Then we do the pivot directly as part of the `from` implementation for only the fields that are returned.

Does that sound like it would work?

Reviewer 2: It seems doable.  I don't think we have any fields vs tags logic in the read path in storage currently, at least for Flux.  It seems like we'll need to change our interface with storage significantly to make this work correctly and efficiently.

# New Flux

Author: Nathaniel Cook

Recently Paul Dix and I added a document on the vision of a new Flux that is more usable.
The design there solves several key points of feedback we have received on Flux.

> That document describe the what and some of the why. This document will discuss the how.

Reviewer: ```suggestion
That document describes the what and some of the why. This document will discuss the how.
```

This document is not meant to be prescriptive rather just my personal commentary on how I see the vision of new Flux could be implemented.
It is meant to be a seed to start a good team discussion around how  we can implement the new Flux vision.

## What is new?

First before we talk about how we would implement the new vision, we should clarify what is new.

### Single Table vs Stream of Tables

The new Flux data model changes to be a single table with a homogeneous schema.
This is different from the current design which represents data as a stream (unbounded sequence) of tables with possibility differing schemas.

> 

Reviewer: This eliminates group keys in favor of a schema? How does this change the way that users are able to group and chart data for comparison?

Author: User's can still group their data, but instead of that grouping being persisted across different functions/transformation the grouping only _lasts_ as long at the function that used it. For example the `aggreate` function will group the data, compute the aggregate of each result and then output a single table with the results from each group. After the call to `aggregate` the grouping is no longer relevant/important.

> This change is fundamental to the new design, the most common pit fall with current Flux is working with this stream of tables that may or may not be the same structure.

Reviewer: This is a current pitfall of Flux? I'm not sure I understand the point made here.

Author: The ways this pitfall manifests is users write a Flux script they expect to work for data following a  certain shape. But the data shape isn't clear from just the script text so they often find the data has a different shape than they expected and so their script doesn't work. To make it even harder the shape of the data can change for each table, so while the script works for some of their data it doesn't work for all of it.

So given that a user needs to understand the shape of their data in order to write a good Flux script, the current design can change the shape of the data in unexpected ways. For example this query changes the group key three different times. Can you spot them?

```
from(bucket:"telegraf") 
    |> range(start: -1h)
    |> filter(fn: (r) => r.env == "acc")
    |> group(columns:["region", "env", "_measurement", "_field"])
    |> mean()
    |> map(fn: (r) => ({region: r.region, env: r.env, _measurement: r._measurement, _value: r._value * 2.0}))
    |> keep(columns:["region", "_value"])
    |> sum()
```

The calls to `group`, `map` and `keep` all change the shape of the data because they change the group key.

The end result is the user has to remember what there group key is and they have to follow the flow of the group key through their script in order to know what it is.

### Influxdb.From is always pivoted

> To go along with the single table and homogeneuos schema the data returned from the `influxdb.from` function will be _pivoted_.

Reviewer: ```suggestion
To go along with the single table and homogeneous schema the data returned from the `influxdb.from` function will be _pivoted_.
```

Specifically the _\_field_ and _\_value_ structure of the data has been removed and instead column names correspond with the name of the InfluxDB tag or field.

> Like before no distinction is made between tags and fields as columns in the data model.

Reviewer: ```suggestion
Like before, no distinction is made between tags and fields as columns in the data model.
```

A consequence of this change is that the `from` call now also requires the `measurement` parameter as the InfluxDB data model only guarantees a homogeneuos schema within the scope of a measurement.

### Syntax Sugar for From

An additional syntax is proposed that makes querying data from InfluxDB much easier to type.

The syntax is:

```
@bucket.measurement

> ```

Reviewer: I really like the simplicity of this syntax. I feel like it will be much for familiar to user's coming from other scripting languages.

The syntax is equivalent to:

```
influxdb.from(bucket:"bucket", measurement:"measurement")

> ```

Author: Yes this is correct. I'd like the syntax to be more generally applicable. Open to ideas there.

> ```

Reviewer: I didn't even pick up on this idea @rockstar, but its an important consideration.

### Syntax Sugar for Filtering

There is new syntax to make creating filter predicates easier. This syntax has two parts.

1. Declares which columns to retrieve from the database.
2. Declares a predicate for which data to retrieve from the database.

The syntax is:

```
{location,state, state == "CA"}
```

That syntax is equivalent to:

```
filter(fn: (r) => r.state == "CA")
```

> With the addition of providing explicit expectation of which columns should be returned from the database.

Reviewer: ```suggestion
This has the addition of providing explicit expectation of which columns should be returned from the database.
```
I wasn't sure of the edit here.

### Syntax Sugar for Aggregates

Aggregates also come with their own new syntax. The intent of this syntax is to make it easy to describe which aggregate function to apply to which columns.
This syntax allows for that definition to be explicit and terse.

The syntax is:

```
 |> aggregate({min(degrees_bottom), max(degrees_bottom), mean(degrees_bottom)}, by:["location"])
```

The semantics of this syntax are to apply three different aggregate functions to the same column named `degrees_bottom`.
The output table will then have three columns named `min_degrees_bottom`, `max_degrees_bottom`, and `mean_degrees_bottom`.

>NOTE: Selectors do not need their own syntax as it is easy to specify selection in the existing syntax.

### Column Indexing

A consequence of data being represented as a single table the `tableFind` function is not needed. As such  a user can call `getColumn` directly on a table.
We can go one step further and use the `.` syntax to mean indexing a table for its column.

For example:

```
data = @foo.bar{location, state == "CA"}

data.location // a list of locations in the state of CA
```

## The How

> Now that we have an overview of the new features syntax lets discuss how I expect they can be implemented.

Reviewer: ```suggestion
Now that we have an overview of the new features syntax, let's discuss how I expect they can be implemented.
```

Again this is just to seed the conversation and is by no means a prescription on how it must be done.

### Single Table vs Stream of Tables

The current Flux internals use the interpreter to create DAG of promises to data to construct a query spec.
That spec is then planned etc and passed down to the execution engine.

> So long as we can continue to build a query spec from this new Flux API we can continue to execute the queries.

Reviewer: ```suggestion
So long as we can continue to build a query spec from this new Flux API, we can continue to execute the queries.
```

My rough plan is that a set of new Flux functions will be added to a `table` package.

> The `table` package will be very similar to the current `universe` package, it will contain all the basic functions and the new functions as well like `aggregate`.

Reviewer: ```suggestion
The `table` package will be very similar to the current `universe` package. It will contain all basic functions and new functions as well like `aggregate`.
```

This new `table` package will need to provide its own mechanism for producing a query spec.
The current execution engine can handle streams of tables with heterogenous schemas.

> A homogenuos schema is a subset of functionality in a heterogeneuos schema and as such can run on the exectution engine without issue.

Reviewer: ```suggestion
A homogenous schema is a subset of functionality in a heterogeneous schema and as such can run on the execution engine without issue.
```

We will still leverage the engines concept of groups to do aggregates etc but those details will be hidden from the user.

> As a rough example to illustrate how this would work the following new Flux:

Reviewer: ```suggestion
As a rough example to illustrate how this would work, the following new Flux:
```

```
@data.temperatures{location,state,degrees_bottom}
 |> aggregate({min(degrees_bottom)}, by:["location"])
```

would generate a spec similar to what this Flux would generate

```
from(bucket:"data")
 |> filter(fn:(r) => r._measurement == "temperatures")
 |> group(columns:["location"])
 |> min()
 |> group()
```

### Influxdb.From is always pivoted

The `influxdb.from` function will be updated to accept a `measurement` parameter.
If it exists the spec will produce the appropriate filter.
If it doesn't exist the behavior will not change.
Once we have versioning we can then break the API of the function and require the measurement parameter.

### Syntax Sugar

Syntax sugar will be implemented as a kind of macro.

> By macro I mean a source to source transformation before any semantic analysis is performed.

Reviewer: Are these `macros` going to be written in pure Flux?

Author: I don't think so, but I don't have a clear idea of how they would be implemented.

Reviewer: I think the idea of framing this syntactic sugar as macros is potentially powerful. It prefers the versatility of the underlying language while providing a stage for more terseness and innovation. I also thinks this may potentially make this language more open source/community friendly depending on how we implement this.

```
(view) => ({
    min_bottom_degrees: min(view.bottom_degrees),
    max_bottom_degrees: max(view.bottom_degrees),
    mean_bottom_degrees: mean(view.bottom_degrees),
})
```

All together then the single aggregate function call expands like this:

```
 |> aggregate({min(degrees_bottom), max(degrees_bottom), mean(degrees_bottom)}, by:["location"])
 // exapands to 
 |> aggregate(
        (view) => ({
            min_bottom_degrees: min(view.bottom_degrees),
            max_bottom_degrees: max(view.bottom_degrees),
            mean_bottom_degrees: mean(view.bottom_degrees),
        }),
        by:["location"]
    )
```

Note we are assuming a positional parameter for the `view` function.

Also notice that the `min`, `max` and `mean` functions now have different type signatures.
Instead of being functions that operate on whole tables, they are functions that operate on list (i.e. columns)

Proposed type signature for the aggregate functions.

```
min : forall [t0] where t0 : Comparable (data: [t0]) -> t0
max : forall [t0] where t0 : Comparable (data: [t0]) -> t0
mean : forall [t0] where t0 : Numeric (data: [t0]) -> float
```

> Finally there is one piece missing from this design, the group columns got dropped.

Reviewer: ```suggestion
Finally, there is one piece missing from this design. The group columns got dropped.
```

In the example you would expect the output data to contain the `location` column, but as I have described the design it would be dropped.
Obviously we need a fix for that, we could play more macro tricks or something. Looking for ideas on this part of the problem.

## Summary

> In summary we see these changes as being additive to the existing language with the addition of some new core features like macros and a new `table` package for these new function types.

Reviewer: ```suggestion
In summary, we see these changes as being additive to the existing language with the addition of some new core features like macros and a new `table` package for these new function types.
```
